File: controlplane/internal/types/task.go
package types

// Task represents an ECS task
type Task struct {
	TaskArn              string            `json:"taskArn,omitempty"`
	ClusterArn           string            `json:"clusterArn,omitempty"`
	TaskDefinitionArn    string            `json:"taskDefinitionArn,omitempty"`
	ContainerInstanceArn string            `json:"containerInstanceArn,omitempty"`
	OverrideId           string            `json:"overrideId,omitempty"`
	LastStatus           string            `json:"lastStatus,omitempty"`
	DesiredStatus        string            `json:"desiredStatus,omitempty"`
	Cpu                  string            `json:"cpu,omitempty"`
	Memory               string            `json:"memory,omitempty"`
	Containers           []Container       `json:"containers,omitempty"`
	StartedBy            string            `json:"startedBy,omitempty"`
	Version              int               `json:"version,omitempty"`
	StoppedReason        string            `json:"stoppedReason,omitempty"`
	Connectivity         string            `json:"connectivity,omitempty"`
	ConnectivityAt       string            `json:"connectivityAt,omitempty"`
	PullStartedAt        string            `json:"pullStartedAt,omitempty"`
	PullStoppedAt        string            `json:"pullStoppedAt,omitempty"`
	ExecutionStoppedAt   string            `json:"executionStoppedAt,omitempty"`
	CreatedAt            string            `json:"createdAt,omitempty"`
	StartedAt            string            `json:"startedAt,omitempty"`
	StoppingAt           string            `json:"stoppingAt,omitempty"`
	StoppedAt            string            `json:"stoppedAt,omitempty"`
	Group                string            `json:"group,omitempty"`
	LaunchType           string            `json:"launchType,omitempty"`
	CapacityProviderName string            `json:"capacityProviderName,omitempty"`
	PlatformVersion      string            `json:"platformVersion,omitempty"`
	PlatformFamily       string            `json:"platformFamily,omitempty"`
	Attachments          []Attachment      `json:"attachments,omitempty"`
	Attributes           []Attribute       `json:"attributes,omitempty"`
	Tags                 []Tag             `json:"tags,omitempty"`
	EnableExecuteCommand bool              `json:"enableExecuteCommand,omitempty"`
	Overrides            *TaskOverride     `json:"overrides,omitempty"`
	EphemeralStorage     *EphemeralStorage `json:"ephemeralStorage,omitempty"`
	HealthStatus         string            `json:"healthStatus,omitempty"`
}

// Container represents a container in a task
type Container struct {
	ContainerArn      string             `json:"containerArn,omitempty"`
	TaskArn           string             `json:"taskArn,omitempty"`
	Name              string             `json:"name,omitempty"`
	Image             string             `json:"image,omitempty"`
	ImageDigest       string             `json:"imageDigest,omitempty"`
	RuntimeId         string             `json:"runtimeId,omitempty"`
	LastStatus        string             `json:"lastStatus,omitempty"`
	ExitCode          int                `json:"exitCode,omitempty"`
	Reason            string             `json:"reason,omitempty"`
	NetworkBindings   []NetworkBinding   `json:"networkBindings,omitempty"`
	NetworkInterfaces []NetworkInterface `json:"networkInterfaces,omitempty"`
	HealthStatus      string             `json:"healthStatus,omitempty"`
	Cpu               string             `json:"cpu,omitempty"`
	Memory            string             `json:"memory,omitempty"`
	MemoryReservation string             `json:"memoryReservation,omitempty"`
	GpuIds            []string           `json:"gpuIds,omitempty"`
}

// NetworkBinding represents a network binding for a container
type NetworkBinding struct {
	BindIP        string `json:"bindIP,omitempty"`
	ContainerPort int    `json:"containerPort,omitempty"`
	HostPort      int    `json:"hostPort,omitempty"`
	Protocol      string `json:"protocol,omitempty"`
}

// NetworkInterface represents a network interface for a container
type NetworkInterface struct {
	AttachmentId       string `json:"attachmentId,omitempty"`
	PrivateIpv4Address string `json:"privateIpv4Address,omitempty"`
	Ipv6Address        string `json:"ipv6Address,omitempty"`
}

// TaskOverride represents overrides for a task
type TaskOverride struct {
	ContainerOverrides []ContainerOverride `json:"containerOverrides,omitempty"`
	TaskRoleArn        *string             `json:"taskRoleArn,omitempty"`
	ExecutionRoleArn   *string             `json:"executionRoleArn,omitempty"`
	Cpu                *string             `json:"cpu,omitempty"`
	Memory             *string             `json:"memory,omitempty"`
	EphemeralStorage   *EphemeralStorage   `json:"ephemeralStorage,omitempty"`
}

// ContainerOverride represents an override for a container
type ContainerOverride struct {
	Name                 *string                `json:"name"`
	Command              []string               `json:"command,omitempty"`
	Environment          []KeyValuePair         `json:"environment,omitempty"`
	EnvironmentFiles     []EnvironmentFile      `json:"environmentFiles,omitempty"`
	Cpu                  *int                   `json:"cpu,omitempty"`
	Memory               *int                   `json:"memory,omitempty"`
	MemoryReservation    *int                   `json:"memoryReservation,omitempty"`
	ResourceRequirements []ResourceRequirement  `json:"resourceRequirements,omitempty"`
}

// RunTaskRequest represents the request to run a task
type RunTaskRequest struct {
	Cluster                  *string                   `json:"cluster,omitempty"`
	TaskDefinition           *string                   `json:"taskDefinition"`
	Count                    *int                      `json:"count,omitempty"`
	Group                    *string                   `json:"group,omitempty"`
	StartedBy                *string                   `json:"startedBy,omitempty"`
	LaunchType               *string                   `json:"launchType,omitempty"`
	CapacityProviderStrategy []CapacityStrategy        `json:"capacityProviderStrategy,omitempty"`
	PlacementConstraints     []PlacementConstraint     `json:"placementConstraints,omitempty"`
	PlacementStrategy        []PlacementStrategy       `json:"placementStrategy,omitempty"`
	PlatformVersion          *string                   `json:"platformVersion,omitempty"`
	EnableECSManagedTags     *bool                     `json:"enableECSManagedTags,omitempty"`
	PropagateTags            *string                   `json:"propagateTags,omitempty"`
	ReferenceId              *string                   `json:"referenceId,omitempty"`
	Tags                     []Tag                     `json:"tags,omitempty"`
	EnableExecuteCommand     *bool                     `json:"enableExecuteCommand,omitempty"`
	Overrides                *TaskOverride             `json:"overrides,omitempty"`
}

// PlacementStrategy represents a placement strategy for a task
type PlacementStrategy struct {
	Type  *string `json:"type"`
	Field *string `json:"field,omitempty"`
}

// PlacementConstraint represents a placement constraint
type PlacementConstraint struct {
	Type       *string `json:"type,omitempty"`
	Expression *string `json:"expression,omitempty"`
}

// RunTaskResponse represents the response from running a task
type RunTaskResponse struct {
	Tasks    []Task    `json:"tasks"`
	Failures []Failure `json:"failures,omitempty"`
}

// StartTaskRequest represents the request to start a task
type StartTaskRequest struct {
	Cluster              *string       `json:"cluster,omitempty"`
	TaskDefinition       *string       `json:"taskDefinition"`
	ContainerInstances   []string      `json:"containerInstances"`
	Group                *string       `json:"group,omitempty"`
	StartedBy            *string       `json:"startedBy,omitempty"`
	EnableECSManagedTags *bool         `json:"enableECSManagedTags,omitempty"`
	PropagateTags        *string       `json:"propagateTags,omitempty"`
	ReferenceId          *string       `json:"referenceId,omitempty"`
	Tags                 []Tag         `json:"tags,omitempty"`
	EnableExecuteCommand *bool         `json:"enableExecuteCommand,omitempty"`
	Overrides            *TaskOverride `json:"overrides,omitempty"`
}

// StartTaskResponse represents the response from starting a task
type StartTaskResponse struct {
	Tasks    []Task    `json:"tasks"`
	Failures []Failure `json:"failures,omitempty"`
}

// StopTaskRequest represents the request to stop a task
type StopTaskRequest struct {
	Cluster *string `json:"cluster,omitempty"`
	Task    *string `json:"task"`
	Reason  *string `json:"reason,omitempty"`
}

// StopTaskResponse represents the response from stopping a task
type StopTaskResponse struct {
	Task *Task `json:"task"`
}

// DescribeTasksRequest represents the request to describe tasks
type DescribeTasksRequest struct {
	Cluster *string  `json:"cluster,omitempty"`
	Tasks   []string `json:"tasks"`
	Include []string `json:"include,omitempty"`
}

// DescribeTasksResponse represents the response from describing tasks
type DescribeTasksResponse struct {
	Tasks    []Task    `json:"tasks"`
	Failures []Failure `json:"failures,omitempty"`
}

// ListTasksRequest represents the request to list tasks
type ListTasksRequest struct {
	Cluster           *string `json:"cluster,omitempty"`
	ContainerInstance *string `json:"containerInstance,omitempty"`
	Family            *string `json:"family,omitempty"`
	StartedBy         *string `json:"startedBy,omitempty"`
	ServiceName       *string `json:"serviceName,omitempty"`
	DesiredStatus     *string `json:"desiredStatus,omitempty"`
	LaunchType        *string `json:"launchType,omitempty"`
	MaxResults        *int    `json:"maxResults,omitempty"`
	NextToken         *string `json:"nextToken,omitempty"`
}

// ListTasksResponse represents the response from listing tasks
type ListTasksResponse struct {
	TaskArns  []string `json:"taskArns"`
	NextToken *string  `json:"nextToken,omitempty"`
}

// GetTaskProtectionRequest represents the request to get task protection
type GetTaskProtectionRequest struct {
	Cluster *string  `json:"cluster,omitempty"`
	Tasks   []string `json:"tasks"`
}

// ProtectionResult represents a protection result for a task
type ProtectionResult struct {
	TaskArn               *string `json:"taskArn"`
	ProtectionEnabled     *bool   `json:"protectionEnabled"`
	ExpirationDate        *string `json:"expirationDate,omitempty"`
	ProtectedFromScaleIn  *bool   `json:"protectedFromScaleIn,omitempty"`
	ProtectedFromScaleOut *bool   `json:"protectedFromScaleOut,omitempty"`
}

// GetTaskProtectionResponse represents the response from getting task protection
type GetTaskProtectionResponse struct {
	ProtectedTasks []ProtectionResult `json:"protectedTasks"`
	Failures       []Failure          `json:"failures,omitempty"`
}

File: web-ui/src/types/task.ts
// Task-related type definitions

export type TaskStatus = 
  | 'PROVISIONING'
  | 'PENDING'
  | 'ACTIVATING'
  | 'RUNNING'
  | 'DEACTIVATING'
  | 'STOPPING'
  | 'DEPROVISIONING'
  | 'STOPPED'
  | 'DELETED';

export type LaunchType = 'EC2' | 'FARGATE' | 'EXTERNAL';

export type HealthStatus = 'HEALTHY' | 'UNHEALTHY' | 'UNKNOWN';

export type Connectivity = 'CONNECTED' | 'DISCONNECTED';

export interface Task {
  taskArn: string;
  clusterArn: string;
  taskDefinitionArn: string;
  capacityProviderName?: string;
  connectivity?: Connectivity;
  connectivityAt?: string;
  containerInstanceArn?: string;
  containers?: Container[];
  cpu?: string;
  createdAt?: string;
  desiredStatus: TaskStatus;
  enableExecuteCommand?: boolean;
  ephemeralStorage?: EphemeralStorage;
  executionStoppedAt?: string;
  group?: string;
  healthStatus?: HealthStatus;
  inferenceAccelerators?: InferenceAccelerator[];
  lastStatus: TaskStatus;
  launchType?: LaunchType;
  memory?: string;
  overrides?: TaskOverride;
  platformFamily?: string;
  platformVersion?: string;
  pullStartedAt?: string;
  pullStoppedAt?: string;
  startedAt?: string;
  startedBy?: string;
  stopCode?: string;
  stoppedAt?: string;
  stoppedReason?: string;
  stoppingAt?: string;
  tags?: Tag[];
  version?: number;
  attributes?: Attribute[];
  availabilityZone?: string;
}

export interface Container {
  containerArn?: string;
  taskArn?: string;
  name: string;
  image?: string;
  imageDigest?: string;
  runtimeId?: string;
  lastStatus?: string;
  exitCode?: number;
  reason?: string;
  networkBindings?: NetworkBinding[];
  networkInterfaces?: NetworkInterface[];
  healthStatus?: HealthStatus;
  managedAgents?: ManagedAgent[];
  cpu?: string;
  memory?: string;
  memoryReservation?: string;
  gpuIds?: string[];
}

export interface NetworkBinding {
  bindIP?: string;
  containerPort?: number;
  hostPort?: number;
  protocol?: 'tcp' | 'udp';
  containerPortRange?: string;
  hostPortRange?: string;
}

export interface NetworkInterface {
  attachmentId?: string;
  privateIpv4Address?: string;
  ipv6Address?: string;
}

export interface ManagedAgent {
  lastStartedAt?: string;
  name?: string;
  reason?: string;
  lastStatus?: string;
}

export interface EphemeralStorage {
  sizeInGiB: number;
}

export interface InferenceAccelerator {
  deviceName: string;
  deviceType: string;
}

export interface TaskOverride {
  containerOverrides?: ContainerOverride[];
  cpu?: string;
  ephemeralStorage?: EphemeralStorage;
  executionRoleArn?: string;
  inferenceAcceleratorOverrides?: InferenceAcceleratorOverride[];
  memory?: string;
  taskRoleArn?: string;
}

export interface ContainerOverride {
  name?: string;
  command?: string[];
  environment?: KeyValuePair[];
  environmentFiles?: EnvironmentFile[];
  cpu?: number;
  memory?: number;
  memoryReservation?: number;
  resourceRequirements?: ResourceRequirement[];
}

export interface InferenceAcceleratorOverride {
  deviceName?: string;
  deviceType?: string;
}

export interface KeyValuePair {
  name?: string;
  value?: string;
}

export interface EnvironmentFile {
  value: string;
  type: 's3';
}

export interface ResourceRequirement {
  value: string;
  type: 'GPU' | 'InferenceAccelerator';
}

export interface Tag {
  key?: string;
  value?: string;
}

export interface Attribute {
  name: string;
  value?: string;
  targetType?: string;
  targetId?: string;
}

// Task Definition types
export interface TaskDefinition {
  taskDefinitionArn?: string;
  family: string;
  taskRoleArn?: string;
  executionRoleArn?: string;
  networkMode?: 'bridge' | 'host' | 'awsvpc' | 'none';
  revision?: number;
  volumes?: Volume[];
  status?: 'ACTIVE' | 'INACTIVE' | 'DELETE_IN_PROGRESS';
  requiresAttributes?: Attribute[];
  placementConstraints?: PlacementConstraint[];
  compatibilities?: string[];
  runtimePlatform?: RuntimePlatform;
  requiresCompatibilities?: string[];
  cpu?: string;
  memory?: string;
  pidMode?: 'host' | 'task';
  ipcMode?: 'host' | 'task' | 'none';
  proxyConfiguration?: ProxyConfiguration;
  registeredAt?: string;
  deregisteredAt?: string;
  registeredBy?: string;
  ephemeralStorage?: EphemeralStorage;
  containerDefinitions: ContainerDefinition[];
}

export interface ContainerDefinition {
  name: string;
  image: string;
  repositoryCredentials?: RepositoryCredentials;
  cpu?: number;
  memory?: number;
  memoryReservation?: number;
  links?: string[];
  portMappings?: PortMapping[];
  essential?: boolean;
  entryPoint?: string[];
  command?: string[];
  environment?: KeyValuePair[];
  environmentFiles?: EnvironmentFile[];
  mountPoints?: MountPoint[];
  volumesFrom?: VolumeFrom[];
  linuxParameters?: LinuxParameters;
  secrets?: Secret[];
  dependsOn?: ContainerDependency[];
  startTimeout?: number;
  stopTimeout?: number;
  hostname?: string;
  user?: string;
  workingDirectory?: string;
  disableNetworking?: boolean;
  privileged?: boolean;
  readonlyRootFilesystem?: boolean;
  dnsServers?: string[];
  dnsSearchDomains?: string[];
  extraHosts?: HostEntry[];
  dockerSecurityOptions?: string[];
  interactive?: boolean;
  pseudoTerminal?: boolean;
  dockerLabels?: Record<string, string>;
  ulimits?: Ulimit[];
  logConfiguration?: LogConfiguration;
  healthCheck?: HealthCheck;
  systemControls?: SystemControl[];
  resourceRequirements?: ResourceRequirement[];
  firelensConfiguration?: FirelensConfiguration;
}

export interface RepositoryCredentials {
  credentialsParameter: string;
}

export interface PortMapping {
  containerPort?: number;
  hostPort?: number;
  protocol?: 'tcp' | 'udp';
  name?: string;
  appProtocol?: string;
  containerPortRange?: string;
}

export interface MountPoint {
  sourceVolume?: string;
  containerPath?: string;
  readOnly?: boolean;
}

export interface VolumeFrom {
  sourceContainer?: string;
  readOnly?: boolean;
}

export interface LinuxParameters {
  capabilities?: KernelCapabilities;
  devices?: Device[];
  initProcessEnabled?: boolean;
  sharedMemorySize?: number;
  tmpfs?: Tmpfs[];
  maxSwap?: number;
  swappiness?: number;
}

export interface KernelCapabilities {
  add?: string[];
  drop?: string[];
}

export interface Device {
  hostPath: string;
  containerPath?: string;
  permissions?: string[];
}

export interface Tmpfs {
  containerPath: string;
  size: number;
  mountOptions?: string[];
}

export interface Secret {
  name: string;
  valueFrom: string;
}

export interface ContainerDependency {
  containerName: string;
  condition: 'START' | 'COMPLETE' | 'SUCCESS' | 'HEALTHY';
}

export interface HostEntry {
  hostname: string;
  ipAddress: string;
}

export interface Ulimit {
  name: string;
  softLimit: number;
  hardLimit: number;
}

export interface LogConfiguration {
  logDriver: string;
  options?: Record<string, string>;
  secretOptions?: Secret[];
}

export interface HealthCheck {
  command: string[];
  interval?: number;
  timeout?: number;
  retries?: number;
  startPeriod?: number;
}

export interface SystemControl {
  namespace?: string;
  value?: string;
}

export interface FirelensConfiguration {
  type: 'fluentd' | 'fluentbit';
  options?: Record<string, string>;
}

export interface Volume {
  name?: string;
  host?: HostVolumeProperties;
  dockerVolumeConfiguration?: DockerVolumeConfiguration;
  efsVolumeConfiguration?: EFSVolumeConfiguration;
  fsxWindowsFileServerVolumeConfiguration?: FSxWindowsFileServerVolumeConfiguration;
}

export interface HostVolumeProperties {
  sourcePath?: string;
}

export interface DockerVolumeConfiguration {
  scope?: 'task' | 'shared';
  autoprovision?: boolean;
  driver?: string;
  driverOpts?: Record<string, string>;
  labels?: Record<string, string>;
}

export interface EFSVolumeConfiguration {
  fileSystemId: string;
  rootDirectory?: string;
  transitEncryption?: 'ENABLED' | 'DISABLED';
  transitEncryptionPort?: number;
  authorizationConfig?: EFSAuthorizationConfig;
}

export interface EFSAuthorizationConfig {
  accessPointId?: string;
  iam?: 'ENABLED' | 'DISABLED';
}

export interface FSxWindowsFileServerVolumeConfiguration {
  fileSystemId: string;
  rootDirectory: string;
  authorizationConfig: FSxWindowsFileServerAuthorizationConfig;
}

export interface FSxWindowsFileServerAuthorizationConfig {
  credentialsParameter: string;
  domain: string;
}

export interface PlacementConstraint {
  type?: 'memberOf';
  expression?: string;
}

export interface RuntimePlatform {
  cpuArchitecture?: 'X86_64' | 'ARM64';
  operatingSystemFamily?: 'WINDOWS_SERVER_2019_FULL' | 'WINDOWS_SERVER_2019_CORE' | 'WINDOWS_SERVER_2016_FULL' | 'WINDOWS_SERVER_2004_CORE' | 'WINDOWS_SERVER_2022_CORE' | 'WINDOWS_SERVER_2022_FULL' | 'WINDOWS_SERVER_20H2_CORE' | 'LINUX';
}

export interface ProxyConfiguration {
  type?: 'APPMESH';
  containerName: string;
  properties?: KeyValuePair[];
}